File: fib/src/fib_omp.c
#include <stdio.h>
#include <stdlib.h>
#include <omp.h>

int fib(int n)
{
  int x,y;

  if (n<2) return n;
  //if n too small, sequence is too short and task creation dominates
  if (n<20) return fib(n-1)+fib(n-2);
  
#pragma omp task default(none) shared(x,n)
  x= fib(n-1);
#pragma omp task default(none) shared(y,n)
  y= fib(n-2);
#pragma omp taskwait
  return x+y;
}

int main (int argc, char *argv[])
{
  int N;
  int res;
  double t_start,t_end;
  
  N=44;

  t_start=omp_get_wtime();
#pragma omp parallel default(none) shared(res,N)
  {
#pragma omp single 
    res=fib(N);
  }
  t_end=omp_get_wtime();
  printf("res=%d\n",res);
  printf("Took %g s\n",t_end - t_start);

  return EXIT_SUCCESS;
}

File: fib/src/fib_omp.f90
PROGRAM fib_omp
  USE omp_lib
  IMPLICIT NONE
  INTEGER :: fib
  INTEGER :: N, res
  REAL(kind=8) :: t_start, t_end
  N=44
  t_start = omp_get_wtime()
  !$OMP PARALLEL DEFAULT(none) SHARED(N,res)
  !$OMP SINGLE
  res=fib(N)
  !$OMP END SINGLE
  !$OMP END PARALLEL
  t_end = omp_get_wtime()

  WRITE(*,*) "result = ",res
  WRITE(*,*) "Expended wall clock time = ",t_end-t_start
  
END PROGRAM fib_omp
  
RECURSIVE FUNCTION fib(n) RESULT(res)
  IMPLICIT NONE

  INTEGER, INTENT(IN) :: n
  INTEGER :: res

  INTEGER :: x,y
  
  IF (n<2) THEN
     res=n
  ELSE IF (n<20) THEN
     x = fib(n-1)
     y = fib(n-2)
     res=x+y
  ELSE
     !$OMP TASK DEFAULT(none) SHARED(x,n)
     x = fib(n-1)
     !$OMP END TASK
     !$OMP TASK DEFAULT(none) SHARED(y,n)
     y = fib(n-2)
     !$OMP END TASK
     !$OMP TASKWAIT     
     res=x+y
  END IF
END FUNCTION fib
